utils/fqdn/fqdn.go

```diff
@@ -0,0 +1,123 @@
+package fqdn
+
+import (
+	"net"
+	"os"
+	"strings"
+	"sync"
+)
+
+// defining os and net package functions in their own variables
+// so that we can mock them for unit tests
+var (
+	osHostname    = os.Hostname
+	netLookupIP   = net.LookupIP
+	netLookupAddr = net.LookupAddr
+)
+
+const unknownHostname = "unknown"
+
+// InitializeAlternativeHostname can be used to set an alternative hostname that is being used by `FindFQDN`.
+// The enforcer can use this during startup to provide an alternative value.
+func InitializeAlternativeHostname(hostname string) {
+	if hostname != "" {
+		alternativeHostnameOnce.Do(func() {
+			alternativeHostnameLock.Lock()
+
+			alternativeHostname = hostname
+
+			alternativeHostnameLock.Unlock()
+		})
+	}
+}
+
+func getAlternativeHostname() string {
+	alternativeHostnameLock.RLock()
+	defer alternativeHostnameLock.RUnlock()
+	return alternativeHostname
+}
+
+var (
+	alternativeHostnameOnce = &sync.Once{}
+	alternativeHostnameLock sync.RWMutex
+	alternativeHostname     string
+)
+
+// Find returns fqdn. It uses the following algorithm:
+// First of all, it will return the globally set alternative hostname if it has been initialized with previously with `IntializeAlternativeHostname`
+// If this is not set, it will try to determine the hostname, resolve the hostname to an IP,
+// and based on the hostname it will perform a reverse DNS lookup for the IP.
+// The first entry of the reverse DNS lookup will be returned.
+// If there are any errors during this process, this function will return "unknown".
+// It will never return an empty string.
+func Find() string {
```

Comment:
that's exactly the part where I actually would like to continue to do the same thing that we used to do before. This new function fixes (and therefore changes) the fqdn detection which can lead to inconsistent behaviour with older enforcers which can lead to multiple registrations etc. which we should definitely try to avoid within 3.14

Reply:
i made sure we don't do anything different here.

Reply:
it's not entirely yet. it can still return "unknown".

Before the logic allowed for empty strings to happen (which is exactly what we have seen at comcast and which is why I believe they are asking for that "alternative hostname" feature to begin with. However, we essentially must not change the extraction behaviour here unless the new feature is used.

Please carefully review the logic from before, let's make sure that we don't change it at all.

Reply:
ok my fault... it is doing what we did before